ocaml/xapi/xapi_vm.ml

```diff
@@ -640,20 +640,34 @@ let provision ~__context ~vm =
 			end
 		end)
 
-(** Sets the maximum number of VCPUs for a {b Halted} guest. *)
+(** Sets the maximum number of VCPUs for a {b Halted} guest
+	or a running guest that is a control domain other than dom0 *)
 let set_VCPUs_max ~__context ~self ~value =
-	if Db.VM.get_power_state ~__context ~self <> `Halted
+	if Helpers.is_domain_zero ~__context self then
```

Comment:
This check assertion seems stricter than what is done in xapi_vm_lifecycle.ml. It would be better to move it in there so that it gets taken into account in the allowed_operations.